### PR TITLE
feat: add notification for suspended/activated account

### LINKED ISSUE #17: feat: account notifications
Relates to https://github.com/coder/coder/issues/13328

Create notification_template instances for the following events, and enqueue notifications with those templates:

- [x] https://github.com/coder/coder/pull/14010
- [x] https://github.com/coder/coder/pull/14113
- [x] https://github.com/coder/coder/pull/14367

These two notifications require some additional functionality, and perhaps require RFCs:

- [ ] https://github.com/coder/coder/issues/14232
  ~password reset (receiver: user)~ ~https://github.com/coder/team-coconut/issues/34~
- [ ] ~user email verification (receiver: user)~ https://github.com/coder/team-coconut/issues/35

### PR DESCRIPTION
**Pull request by mtojek**
Fixes: https://github.com/coder/team-coconut/issues/17

This PR extends `api.putUserStatus` to notify owners, user admins, and target accounts when the account is suspended/reactivated.

### CODE REVIEW

File: coderd/database/migrations/000246_notifications_user_suspended.up.sql [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        'Workspace Events', '[
+        {
+            "label": "View accounts",
+            "url": "{{ base_url }}/deployment/users?filter=status%3Aactive"
+        }
+    ]'::jsonb);
+INSERT INTO notification_templates (id, name, title_template, body_template, "group", actions)
+VALUES ('9f5af851-8408-4e73-a7a1-c6502ba46689', 'User account activated', E'User account "{{.Labels.activated_account_name}}" activated',
+        E'Hi {{.UserName}},\nUser account **{{.Labels.activated_account_name}}** has been activated.',
+        'Workspace Events', '[
```
Comment (dannykopping) [inline coderd/database/migrations/000246_notifications_user_suspended.up.sql:13]:
```suggestion
        'User Events', '[
        {
            "label": "View accounts",
            "url": "{{ base_url }}/deployment/users?filter=status%3Aactive"
        }
    ]'::jsonb);
INSERT INTO notification_templates (id, name, title_template, body_template, "group", actions)
VALUES ('9f5af851-8408-4e73-a7a1-c6502ba46689', 'User account activated', E'User account "{{.Labels.activated_account_name}}" activated',
        E'Hi {{.UserName}},\nUser account **{{.Labels.activated_account_name}}** has been activated.',
        'User Events', '[
```

File: coderd/database/migrations/000246_notifications_user_suspended.up.sql [hunk 2]
```diff
@@ -0,0 +1,18 @@
+INSERT INTO notification_templates (id, name, title_template, body_template, "group", actions)
+VALUES ('b02ddd82-4733-4d02-a2d7-c36f3598997d', 'User account suspended', E'User account "{{.Labels.suspended_account_name}}" suspended',
+        E'Hi {{.UserName}},\nUser account **{{.Labels.suspended_account_name}}** has been suspended.',
+        'Workspace Events', '[
+        {
+            "label": "View accounts",
```
Comment (dannykopping) [inline coderd/database/migrations/000246_notifications_user_suspended.up.sql:6]: Shouldn't this link to suspended accounts?

File: coderd/users.go [hunk 3]
```diff
@@ -857,8 +857,61 @@ func (api *API) putUserStatus(status database.UserStatus) func(rw http.ResponseW
 			})
 			return
 		}
-		aReq.New = suspendedUser
+		aReq.New = targetUser
+
+		// Notify about the change of user status
```
Comment (dannykopping) [inline coderd/users.go:862]: Nit: extracting this to a function would probably read better.

File: coderd/users.go [hunk 4]
```diff
@@ -857,8 +857,61 @@ func (api *API) putUserStatus(status database.UserStatus) func(rw http.ResponseW
...
+		userAdmins, err := api.Database.GetUsers(ctx, database.GetUsersParams{
+			RbacRole: []string{codersdk.RoleUserAdmin},
+		})
+		if err != nil {
+			httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+				Message: "Internal error fetching user admins",
+				Detail:  err.Error(),
+			})
+			return
+		}
```
Comment (dannykopping) [inline coderd/users.go:900]: Replace with `findUserAdmins`.

File: coderd/users.go [hunk 5]
```diff
@@ -857,8 +857,61 @@ func (api *API) putUserStatus(status database.UserStatus) func(rw http.ResponseW
...
+
+		// Send notifications to user admins and affected user
```
Comment (dannykopping) [inline coderd/users.go:902]: If we're addressing the affected user, it'll be a bit weird to receive a notification referring to their account in the 3rd person.

`User "bobby" has been suspended` vs `Your account has been suspended`.

There will be different actions to be taken by the different personas, as well.
I think we need to create a new template for this case.

File: coderd/users.go (hunk 4 above)
Comment (mtojek) [reply coderd/users.go]: 🤦‍♂️

File: coderd/users.go (hunk 5 above)
Comment (mtojek) [reply coderd/users.go]: Fixed

File: coderd/users.go [hunk 6]
```diff
@@ -857,7 +857,16 @@ func (api *API) putUserStatus(status database.UserStatus) func(rw http.ResponseW
 			})
 			return
 		}
-		aReq.New = suspendedUser
+		aReq.New = targetUser
+
+		err = api.notifyUserStatusChanged(ctx, user, status)
+		if err != nil {
+			httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
```
Comment (dannykopping) [inline coderd/users.go:864]: I don't think we should fail the HTTP request; notifications are merely advisory.

File: coderd/users.go [hunk 7]
```diff
@@ -867,9 +876,52 @@ func (api *API) putUserStatus(status database.UserStatus) func(rw http.ResponseW
...
+		adminTemplateID = notifications.TemplateUserAccountActivated
+		ownerTemplateID = notifications.TemplateYourAccountActivated
+	default:
+		api.Logger.Error(ctx, "user status is not supported", slog.F("username", user.Username), slog.F("user_status", string(status)))
+		return xerrors.Errorf("unable to notify admins as the user's status is unsupported")
+	}
+
+	userAdmins, err := findUserAdmins(ctx, api.Database)
+	if err != nil {
+		return xerrors.Errorf("unable to find user admins: %w", err)
```
Comment (dannykopping) [inline coderd/users.go:902]: We should still be able to send notifications to the affected user though even if this occurs, no?

File: coderd/users.go [hunk 8]
```diff
@@ -867,9 +876,52 @@ func (api *API) putUserStatus(status database.UserStatus) func(rw http.ResponseW
...
+}
+
+func (api *API) notifyUserStatusChanged(ctx context.Context, user database.User, status database.UserStatus) error {
+	var key string
+	var adminTemplateID, ownerTemplateID uuid.UUID
+	switch status {
+	case database.UserStatusSuspended:
+		key = "suspended_account_name"
+		adminTemplateID = notifications.TemplateUserAccountSuspended
+		ownerTemplateID = notifications.TemplateYourAccountSuspended
```
Comment (dannykopping) [inline coderd/users.go:890]: Nit: `owner` is a bit of a strange name; perhaps `userTemplateID` or `personalTemplateID`?

**Review by dannykopping — APPROVED** [review]
LGTM, nice refactor.
A couple minor things to address, but I don't need to review again.

**Pull request merged**